sycl/include/CL/sycl/access/access.hpp

```diff
@@ -187,17 +187,15 @@ template <typename ElementType>
 struct DecoratedType<ElementType, access::address_space::local_space> {
   using type = __OPENCL_LOCAL_AS__ ElementType;
 };
+template <class T> struct remove_AS { typedef T type; };
 
-template <class T>
-struct remove_AS {
-  typedef T type;
+template <class T> struct deduce_AS {
+  static const access::address_space value =
+      access::address_space::global_space;
 };
 
 #ifdef __SYCL_DEVICE_ONLY__
-template <class T>
-struct remove_AS<__OPENCL_GLOBAL_AS__ T> {
-  typedef T type;
-};
+template <class T> struct remove_AS<__OPENCL_GLOBAL_AS__ T> { typedef T type; };
```

Comment:
`deduce_AS<__OPENCL_GLOBAL_AS__ T>` is missing.

Please, add a test case to detect this issue.

Reply:
In fact deduce_AS<__OPENCL_GLOBAL_AS__ T> goes to L192 and gets expected result.

Reply:
> The address space information is deduced automatically. Only pointers to global and local address spaces are valid. Passing a pointer to private address space will cause an assertion. **Other address spaces are cast to global with potentially undefined behavior.**

According to my understanding address space deduction is done by device compilers. Why do we apply an exception for sub-group load/store?

Reply:
Good, but I have concerns regarding new "deduction" approach implemented in templates. Assuming "global" by default might cause potential issues. It would be great to avoid them.

Reply:
In the latest patches only local/global pointers are supported. For others assertion is generated (document has been updated accordingly).

The deduction here is extracting AS information from type attribute to sycl::access::address_space enumeration. Renaming  suggestions are welcome.

Set default value to undefined to avoid compilation with unsupported address space.